pubnub-core/src/subscription/subscription.rs

```diff
@@ -0,0 +1,95 @@
+use super::subscribe_loop::*;
+use crate::message::Message;
+use crate::runtime::Runtime;
+use futures_channel::mpsc;
+use futures_util::sink::SinkExt;
+use futures_util::stream::Stream;
+use futures_util::task::{Context, Poll};
+use log::debug;
+use std::pin::Pin;
+
+/// # PubNub Subscription
+///
+/// This is the message stream returned by [`PubNub::subscribe`]. The stream yields [`Message`]
+/// items until it is dropped.
+///
+/// [`PubNub::subscribe`]: crate::pubnub::PubNub::subscribe
+#[derive(Debug)]
+pub struct Subscription<TRuntime: Runtime> {
+    pub(crate) runtime: TRuntime,  // Runtime to use for managing resources
+    pub(crate) name: ListenerType, // Channel or Group name
+    pub(crate) id: SubscriptionID, // Unique identifier for the listener
+    pub(crate) control_tx: ControlTx, // For cleaning up resources at the subscribe loop when dropped
+    pub(crate) channel_rx: ChannelRx, // Stream that produces messages
+}
+
+/// `Subscription` is a stream.
+impl<TRuntime: Runtime> Stream for Subscription<TRuntime> {
+    type Item = Message;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        Stream::poll_next(Pin::new(&mut self.get_mut().channel_rx), cx)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        Stream::size_hint(&self.channel_rx)
+    }
+}
+
+impl<TRuntime: Runtime> Subscription<TRuntime> {
+    /// Request to unsubscribe the subscription from the channel.
+    ///
+    /// This call completes when the unsubscription request is submitted
+    /// successfully, but doesn't indicate when the unsubscription request
+    /// is processed by the subscribe loop control command handler.
```

Comment:
Why is this necessary? `std::mem::drop(subscription);` does the same thing.

Reply:
With `drop` you can't wait for the actual drop to happen. From the use case perspective, this might be necessary to ensure no network requests are issued by the subscribe loop after a certain point. Using `drop ` would not be sufficient since it does not wait for the command to complete.

Having both `Drop` impl and this fn covers both resource leakage prevention when user doesn't care about the precise order of the events (i.e. doesn't need to ensure the subscribe loop processes drop before the user code continues), and the case when it matters for the user.

A few days ago, I was thinking about this implementation and I've spotted an issue. The `Subscription` struct should track and only send a single unsubscription request. I deem this should be implemented by a better abstraction - like a consumable one-shot unsubscription handle for `Subscription` to own, atop of the mpsc control queue.

Reply:
There might be a use case for waiting for the network layer to do its thing, but I haven't ever found one in practice. Perhaps this is more important in resource-constrained environments?

Once dropped, messages can no longer be received on that channel, even if the network layer (and subscription manager) manage to receive anything after the drop. This case is gracefully handled, and the end user doesn't need to be aware of it.

A big part of the challenge is actually in adapting a streaming interface over a multiplexed HTTP-based protocol. After working with this specific protocol for nearly a decade, I'm pretty confident that we really _don't_ want to allow unfortunate parts of this protocol abstraction to leak into the streaming interface.

The other point I want to make is that Rust doesn't have an asynchronous `Drop`, and it might never happen. But the need for it has at least been identified: https://boats.gitlab.io/blog/post/poll-drop/ see also discussion on async cancelation, which is very similar in concept and necessity: https://internals.rust-lang.org/t/async-await-the-challenges-besides-syntax-cancellation/10287 (the linked gist has several relevant details).

Assuming something like either of these could be relied upon, I would lean heavily toward just dropping the stream when a subscription needs to be, quite literally, dropped. At least for now, since it's the model that most closely fits with existing stream mechanisms in the Rust ecosystem. This might change over time as the ecosystem evolves.

Reply:
I'm actually using in the tests already to orchestrate the behavior of the control loop into edge cases. This kind of workaround is pretty common, it is indeed specific to the async drop absence.

There are two ways we could implement drop: with blocking internal cleanup operation, and with non-blocking, but offloaded to a separately scheduled `Task`. The first approach makes it impossible to use the implementation in the async context because blocking would be critically problematic for the whole runtime. So, the second way is typically used, but it has another problem - you can't wait for the actual resource cleanup completion. For this reason, async `Drop` implementers typically also provide the async function that returns a future which resolves after the resource allocation if fully done (or similar reasonable logical even is reached).

Practically, the cancellation is already effectively solved by the way we handle commands in the control loop, so all we need is to just send the drop command. From the interface point of view, the `Subscription` struct is our cancellation token, so we can drop it (or issue an unsubscription in an async-firedly way as implemented currently).

Actually, it may be problematic to have an async operations in the drop in the first place. A trivial (and unoptiomal - more on this later) way of walking around it would be to pass the drop requests not via the typical control tx we usually use, but via an `oneshot` channel, sending over which is not async. This would require maintaining a `oneshot` channel per subscription, which has a lot of overhead. If we go with a lower-level implementation, we can make a specialized custom channel to do this optimally: send without blocking, and receive from multiple producers with just a single waker. I have experience with implementing these sorts of things, and will be able to implement one, however it might not be high in the priority list.

To sum up, there are multiple ways we can change the current implementation, and, I added the `unsubscribe` call to actually solve a pain point that wasn't covered otherwise - precise control over the control loop in the tests. We can approach the solution for that pain point differently, and eliminate the very need for this `unsubscribe` call - for example via additional control loop hooks. I expect more options to unravel as we continue refactoring other parts further. For this reason, I'd propose to just postpone the decision on this API call for now - if we remove it now, it'd be troublesome to support some of the test cases (among the most important ones, since the ones in question did actually caught bugs), and if we implement something else and the refactor other parts we might end up changing this again after that cause there will be a better option. So I'd recommend to just keep in mind for now that we might want to get rid of it.

Reply:
> So, the second way is typically used, but it has another problem - you can't wait for the actual resource cleanup completion.

I'm honestly happy with this tradeoff for now. Like I alluded earlier, there is probably a use case for waiting, I just haven't seen it myself in practice. As long as we can guarantee that resource cleanup will be performed when necessary, the cleanliness of the `Drop` approach feels much more natural.

I'm aware that `Drop` cannot always be guaranteed, because of the reference cycle problem (and panic unwinds and other related issues). My argument here is that if this behavior is acceptable to Rust, it should be acceptable to any crate written in Rust. Memory leaks are not memory safety violations.

> Practically, the cancellation is already effectively solved by the way we handle commands in the control loop, so all we need is to just send the drop command. From the interface point of view, the Subscription struct is our cancellation token, so we can drop it ...

Exactly. 

> (or issue an unsubscription in an async-firedly way as implemented currently).

With the challenge that it's a bit of an anti-pattern that will allow a dead `Subscription` to be used in some other logical thread, including whatever error handling may be necessary for handling this new edge case.

A synchronous drop would be awesome if it can be done (it's what I wanted all along). Feels like it will almost certainly have to cross a thread boundary, though. I'm wary of that issue. And that's why `Drop` reuses the `channel` at present.

Reply:
We can definitely improve a way the `unsubscription` is implemented with the current approach, as has a few bugs, as I said above.
If you’re worried about this being an anti-pattern though, I would disagree. And it is not about memory safety or leaks, it is about eliminating race conditions. The similar functionality is commonly provided across the rust ecosystem:
- https://docs.rs/futures/0.3.1/futures/channel/mpsc/struct.Receiver.html#method.close
- https://docs.rs/tokio/0.2.8/tokio/sync/mpsc/struct.Receiver.html#method.close
- similar stuff for sockets and so on

> I'm honestly happy with this tradeoff for now. Like I alluded earlier, there is probably a use case for waiting, I just haven't seen it myself in practice. As long as we can guarantee that resource cleanup will be performed when necessary, the cleanliness of the `Drop` approach feels much more natural.

I added the `unsubscribe` strictly for practical reasons, there are a couple new tests that simply can’t be implemented with just the `Drop`, exactly because it causes a race condition (and in fact don’t produce correct results if `drop(s)` is used instead of `s.unsubscribe().await`.

It is true though that we could make `unsubscribe` take `self` by value - to simplify the interface. Currently, it is still possible to read messages after the unsubscription request was submitted - because there’s a race between when us sending a command to control loop and the http request completion. Technically, we could end up with new messages being available in the subscription, while the subscription has the drop command pending on the control loop. User might want to handle this case properly - and it’s reasonable to let this, since the compute resources have been spent, and we have the meaningful results. The beauty here is it’s the user’s discretion whether he wants to handle that case or not. If one doesn’t, a usual drop will suffice, and the there’s practically no need to use this method for those use cases.

Take a look at this commit: https://github.com/pubnub/rust/commit/254df9672862fdee3742812ab923508f2dfa7e83
There, the main value is the test itself, as it does check some interesting stuff. However, to be able to actually implement it I *had to* add the `unsubscribe` call. So, if we were to change this, we’d have to deal with the test breakage too. The test is important though. It is nothing we can’t solve otherwise somehow, but addressing it a few steps later might be easier, cause we’ll be able to focus solely on this issue with other things out of the door.

Reply:
I must admit, I don't understand the purpose of the linked test. The transport layer should be safe to cancel at any time. I get that there are race conditions, but most race conditions in this code are the "don't care" variety.

Could you provide more details about why it matters that the subscribe loop can be terminated while an HTTP request is outstanding? Or if I've misunderstood the intention, could you clarify the intention of the test?

Reply:
The meaningful part is here: https://github.com/pubnub/rust/commit/254df9672862fdee3742812ab923508f2dfa7e83#diff-6ab62c8d2d967ad2931b0a62a8186475R166-R194

This integration test ensures there're no issues with the subscribe loop termination and restart logic. It doesn't really test the transport (at least that wasn't the intention).

I'm reading that test code now, and apparently, eliminating the `unsubscribe` and just dropping the subscription shouldn't cause any issues. I apologize, I think I've got confused somehow - I was under the impression it was required.

Now, I don't think there's a real internal use case for the `unsubscribe`. Given that it's kind of redundant, I don't mind removing it. We can add it again later if there's a need.

Reply:
Nice! Yeah, I think we are in agreement on this one now.